File: backend/app/models/task.py
from datetime import date, datetime
from typing import TYPE_CHECKING

from sqlalchemy import Date, DateTime, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.database import Base

if TYPE_CHECKING:
    from app.models.tag import Tag


class Task(Base):
    __tablename__ = "tasks"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    title: Mapped[str] = mapped_column(String(120), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(80), nullable=False, default="К выполнению", index=True)
    priority: Mapped[str | None] = mapped_column(String(20), nullable=True, index=True)
    deadline: Mapped[date | None] = mapped_column(Date, nullable=True)
    parent_task_id: Mapped[int | None] = mapped_column(Integer, nullable=True, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )
    tags: Mapped[list["Tag"]] = relationship("Tag", secondary="task_tags", back_populates="tasks")

File: frontend/src/types/task.ts
export type TaskPriority = 'Низкий' | 'Средний' | 'Высокий'
export type TaskStatus = 'К выполнению' | 'В работе' | 'Выполнено'

export type Task = {
  id: number
  title: string
  description: string | null
  status: string
  priority: TaskPriority | null
  deadline: string | null
  parent_task_id: number | null
  tags: { id: number; name: string }[]
  created_at: string
  updated_at: string
}

export type TaskFilters = {
  status?: string
  priority?: TaskPriority
  search?: string
  deadline_from?: string
  deadline_to?: string
  tag?: number
  sort_by?: 'created_at' | 'deadline' | 'priority' | 'status'
  sort_order?: 'asc' | 'desc'
}

export type TaskPayload = {
  title: string
  description?: string | null
  status?: string
  priority?: TaskPriority | null
  deadline?: string | null
  parent_task_id?: number | null
  tag_ids?: number[]
}
